### PR TITLE
Throw a type error when using hk-types in unions or intersections

### LINKED ISSUE #16696: Union and intersection types are not causing checking the kind of type parameters
## Compiler version
3.2.1

## Minimized code

https://scastie.scala-lang.org/VEozYp5OQlmLyTb0hn3KQw

## Output (click arrow to expand)
<details>

```scala
java.lang.AssertionError: assertion failed: R &amp; Playground.Foo / TypeParamRef(R) &amp; HKTypeLambda(List(A), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), AppliedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class &lt;empty&gt;)),module class Playground$)),trait Foo),List(TypeParamRef(A))))
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.core.Types$AndType$.apply(Types.scala:3203)
	at dotty.tools.dotc.core.Types$AndType$.make(Types.scala:3238)
	at dotty.tools.dotc.core.Types$AndType.derivedAndType(Types.scala:3184)
	at dotty.tools.dotc.core.Types$ApproximatingTypeMap.derivedAndType(Types.scala:5961)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5689)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:109)
	at dotty.tools.dotc.core.Types$TypeMap.op$proxy16$1(Types.scala:5590)
	at dotty.tools.dotc.core.Types$TypeMap.mapArgs(Types.scala:5590)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5627)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:109)
	at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5603)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:103)
	at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5603)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:103)
	at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5603)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:103)
	at dotty.tools.dotc.core.TypeOps$.asSeenFrom(TypeOps.scala:51)
	at dotty.tools.dotc.core.Types$Type.asSeenFrom(Types.scala:1036)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.derived$1(Denotations.scala:1099)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1125)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1079)
	at dotty.tools.dotc.core.Denotations$PreDenotation.asSeenFrom(Denotations.scala:134)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.findMember(SymDenotations.scala:2069)
	at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:686)
	at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:879)
	at dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:669)
	at dotty.tools.dotc.core.Types$Type.member(Types.scala:653)
	at dotty.tools.dotc.typer.ProtoTypes$SelectionProto.liftedTree1$1(ProtoTypes.scala:195)
	at dotty.tools.dotc.typer.ProtoTypes$SelectionProto.isMatchedBy(ProtoTypes.scala:217)
	at dotty.tools.dotc.core.TypeComparer.isMatchedByProto(TypeComparer.scala:1917)
	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:314)
	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1336)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:194)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:204)
	at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
	at dotty.tools.dotc.core.TypeComparer.testSubType(TypeComparer.scala:144)
	at dotty.tools.dotc.core.TypeComparer$.testSubType(TypeComparer.scala:2767)
	at dotty.tools.dotc.typer.Typer.adaptNoArgsOther$1(Typer.scala:3725)
	at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:3822)
	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:4038)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3368)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
```
</details>

### PR DESCRIPTION
**Pull request by odersky**
Throw a type error instead of crashing when using higher-kinded types in unions or intersections.

We check kindedness only in PostTyper (to avoid cycles), which means we might get into situations where we combine a higher-kinded type in a union or intersection, which is illegal. In this case we now diagnose the problem with a TypeError instead of failing an assert. However, after Typer we do fail since by then such situations should have been checked by then.

Fixes #16696

### CODE REVIEW

File: tests/new/test.scala [hunk 1]
```diff
@@ -1,2 +1,9 @@
...
-  def test = 0
+trait A
+
+implicit def stringToA(s: String): A = new A{override def toString = s}
+implicit def unitToA(u: Unit): A = new A{override def toString = "Unit"}
+
+val a0 = if (System.nanoTime() != 4) "hello"
+val a: A = a0
+@main def Test =
+  println(a)
```
Comment (dwijnand) [inline tests/new/test.scala:9]: Looks like a commit mistake?

**Review by dwijnand — APPROVED** [review]
LGTM

File: tests/neg/i16696.check [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  |                             ^
+  |                             Some is not a value type, cannot be used in in intersection Some & Int
+-- Error: tests/neg/i16696.scala:8:29 ----------------------------------------------------------------------------------
+8 |  val boom2 = BoxMaker[Some].make2 // error
+  |                             ^
+  |                             Some is not a value type, cannot be used in in union Some | Int
+-- Error: tests/neg/i16696.scala:20:27 ---------------------------------------------------------------------------------
+20 |  val boom = BoxMaker[Foo].make(_.foo) // error
+   |                           ^
+   |                           test2.Foo is not a value type, cannot be used in in intersection R & test2.Foo
```
Comment (dwijnand) [inline tests/neg/i16696.check:12]: actually, it reads "in in"

Comment (odersky) [reply tests/neg/i16696.check]: Ah yes. Good that you spotted! it!

**Pull request merged**